src/physics/em/detail/RayleighInteractor.i.hh

```diff
@@ -15,54 +18,123 @@ namespace detail
  */
 template<class Engine>
 CELER_FUNCTION Interaction RayleighInteractor::operator()(Engine& rng)
 {
-    // Allocate space for XXX (electron, multiple particles, ...)
-    Secondary* secondaries = this->allocate_(0); // XXX
-    if (secondaries == nullptr)
-    {
-        // Failed to allocate space for a secondary
-        return Interaction::from_failure();
-    }
-
-    // XXX sample
-    (void)sizeof(rng);
+    real_type energy = inc_energy_.value();
 
     // Construct interaction for change to primary (incident) particle
     Interaction result;
-    result.action      = Action::scattered;                     // XXX
-    result.energy      = units::MevEnergy{inc_energy_.value()}; // XXX
-    result.direction   = inc_direction_;
-    result.secondaries = {secondaries, 1}; // XXX
+    result.action = Action::scattered;
+    result.energy = units::MevEnergy{inc_energy_.value()};
+
+    // Sample direction for a given Z: G4RayleighAngularGenerator
+    ItemIdT item_id = ItemIdT{element_id_.get()};
+
+    Real3 pn = shared_.params.data_n[item_id];
+    Real3 pb = shared_.params.data_b[item_id];
+    Real3 px = shared_.params.data_x[item_id];
+
+    SampleInput input = this->evaluate_weight_and_prob(energy, pb, pn, px);
+
+    real_type                          cost;
+    UniformRealDistribution<real_type> u01(0, 1.0);
+
+    do
+    {
+        unsigned  index = 0;
+        real_type x     = u01(rng);
+
+        if (x > input.prob[0])
+        {
+            index = (x <= input.prob[0] + input.prob[1]) ? 1 : 2;
+        }
+
+        real_type w = input.weight[index];
+        real_type n = pn[index];
+        real_type b = pb[index];
+
+        n = 1.0 / n;
+
+        // Sampling of scattering angle
+        real_type y = w * u01(rng);
+        if (y < num_limit())
+        {
+            x = y * n * (1. + 0.5 * (n + 1.) * y * (1. - (n + 2.) * y / 3.));
+        }
+        else
+        {
+            x = std::exp(-n * std::log(1. - y)) - 1.0;
+        }
 
-    // Save outgoing secondary data
-    secondaries[0].particle_id = shared_.electron_id; // XXX
-    secondaries[0].energy      = units::MevEnergy{0}; // XXX
-    secondaries[0].direction   = {0, 0, 0};           // XXX
+        cost = 1.0 - 2.0 * x / (b * input.factor);
+
+    } while (2 * u01(rng) > 1.0 + cost * cost || cost < -1.0);
+
+    UniformRealDistribution<real_type> sample_phi(0, 2 * constants::pi);
+
+    // Scattered direction
+    result.direction
+        = rotate(from_spherical(cost, sample_phi(rng)), inc_direction_);
 
     return result;
 }
 
+CELER_FUNCTION
+auto RayleighInteractor::evaluate_weight_and_prob(real_type    energy,
+                                                  const Real3& b,
+                                                  const Real3& n,
+                                                  const Real3& px) const
+    -> SampleInput
+{
+    SampleInput input;
+
+    real_type factor = energy * hc_factor();
+    input.factor     = factor * factor;
```

Comment:
```suggestion
    input.factor = ipow<2>(energy * RayleighInteractor::hc_factor());
```

Reply:
As I see it, it's a matter of telling someone "where do I look to find this function?":
- For local or  in-the-same-file functions, no namespace is needed
- For a free function in a different header, generally use `celeritas::` or `detail::` as appropriate
- For member functions of the class, use `this->`
- For static functions of the class, use `ClassName::` (though I'm OK with aliasing `This = RayleighInteractor` so you can call `This::`)

I don't want to make a big deal about it so if you think readability is *worse* by adding `RayleighInteractor::` we can skip it.

Reply:
Aha! RayleighInteractor:: is used instead of this-> since it is a static member function.  Clear enough!